### ISSUE TITLE
Publish doesn't push tags if using --registry

### ISSUE THREAD

**Issue by nik0kin** [microsoft/web-build-tools#1430]
## Please prefix the issue title with the project name i.e. [rush], [api-extractor] etc. ##

**Is this a feature or a bug?**

- [ ] Feature
- [ ] Bug
- [X] Bug/Confusion

**Please describe the actual behavior.**

Running `rush publish -a -p -b master -r https://private-npm-registry/repository/npm-hosted/` does not push tags, it DRYRUNs for tags only (`* DRYRUN: git tag -a @package_v1.0.12 -m @package v1.0.12`)

Packages are successfully published, the publish commit is pushed, just no tags.

**What is the expected behavior?**

Tags will be published.

It seems https://github.com/microsoft/web-build-tools/blob/4811cea4782cbb7b6a55c7e320e67ebcf5e86236/apps/rush-lib/src/cli/actions/PublishAction.ts#L340 is causing only publishes without registryUrls to push tags.

Am I supposed to push tags in a second publish step?

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** Rush
* **Tool Version:** 5.10.3
* **Node Version:** v8.11.4
* **OS:** MacOS

**Comment by nik0kin** [comment]
What is the logic behind:

```
  -r REGISTRY, --registry REGISTRY
                        Publishes to a specified NPM registry. If this is
                        specified, it will prevent the current commit will
                        not be tagged.
```

This is preventing my publish command from tagging while publishing to a private npm repo.

**Comment by nik0kin** [comment]
It appears that if I remove ` && !this._registryUrl.value` from https://github.com/microsoft/web-build-tools/blob/4811cea4782cbb7b6a55c7e320e67ebcf5e86236/apps/rush-lib/src/cli/actions/PublishAction.ts#L340 everything publishes and tags are pushed correctly.

Would you consider this change?

**Comment by matthiasg** [comment]
@rakeshpatnaik can anybody explain what the reasoning is ?

Also is the registry url even necessary if the individual package.jsons include a `publishConfig.registry` ? That way it could be potentially circumvented.

**Comment by octogonz** [comment]
> Also is the registry url even necessary if the individual package.jsons include a `publishConfig.registry` ? That way it could be potentially circumvented.

This might be awkward to maintain for a repo with hundreds of Rush projects. Also, it's common for the same project to get published to different registries (e.g. to the private registry for internal usage, to a temporary registry for release testing, and to npmjs.com for public releases).

**Comment by octogonz** [comment]
> @rakeshpatnaik can anybody explain what the reasoning is ?

That code snippet calls this function, whose comment seems to explain it:
```ts
  public addTag(shouldExecute: boolean, packageName: string, packageVersion: string, commitId: string | undefined): void {
    // Tagging only happens if we're publishing to real NPM and committing to git.
    const tagName: string = PublishUtilities.createTagname(packageName, packageVersion);
    PublishUtilities.execCommand(
      !!this._targetBranch && shouldExecute,
      'git',
      ['tag', '-a', tagName, '-m', `${packageName} v${packageVersion}`, ...(commitId ? [commitId] : [])]);
  }
```

If we revisit the three examples from above:

1.  to the private registry for internal usage
2.  to a temporary registry for release testing
3.  to npmjs.com for public releases

...I think the reasoning would be that (3) is the important one with meaningful tags. Whereas (1) would happen hourly and its version numbers are not very meaningful, and (2) is for temporary purposes.

So the person who designed this logic (who has not been involved with Rush) seems to have baked some of their own assumptions into the general logic.  The publishing feature was developed very early in Rush's history and is not very easy to understand. We are planning to overhaul it.

To fix your problem, I think we we want is a way to explicitly specify whether tags get created or not, rather than relying on these heuristics about the `registryUrl`.  

Can someone suggest a solution? I don't have much time to think through this problem right now. @nik0kin 's proposal above is not bad, but I'm concerned it might be breaking change for existing CI scripts. Is there a way to make the superior behavior opt-in?

**Comment by sdalonzo** [comment]
👍 for being able to just pass a discrete flag for this `--addGitTags` or similar.

**Comment by danfma** [comment]
Hello, is that solved by `--apply-git-tags-on-pack`? I'm running into this problem right now.

**Comment by danfma** [comment]
> Hello, is that solved by `--apply-git-tags-on-pack`? I'm running into this problem right now.

No, apparently it has nothing to do with that. Is there any workaround to handle this problem?

**Comment by isc30** [comment]
hey, facing this issue, was it fixed somehow?
